drivers: sensors: Add support for Nordic QDEC

pdunaj opened the pull request.

pdunaj committed:
ext: hal: nordic: Enable building of NRFX QDEC

Allow project to use Nordic's NRFX QDEC driver.

This closes issue #9364

Jira:DESK-259

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

pdunaj committed:
dts: Add support for Nordic QDEC

Add QDEC binding and entry in Nordic SoC DTSI files.

This closes issue #9364

Jira:DESK-259

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

pdunaj committed:
include: Add rotation to sensor channels

This channel is aimed to be used by QDECs.

This closes issue #9364

Jira:DESK-259

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

anangl commented on drivers/sensor/qdec_nrfx/qdec_nrfx.c:
```diff
@@ -0,0 +1,306 @@
+#include <hal/nrf_gpio.h>
+
+#define SYS_LOG_DOMAIN "qdec_nrfx"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_QDEC_NRFX_LEVEL
```
Definition of this Kconfig option seems to be missing.
And perhaps it would be better to use the new logger (_<logging/log.h>_) right away.

pdunaj replied:
fixed
as for the new logger, to be honest I would like to leave it as is and commit it faster.
Other drivers are not using it yet so it should be no problem I guess since somebody will have to do a bulk change anyway.

pdunaj commented:
fixed and rebased

pdunaj commented:
there would be one more change related to qdec that I need but I would prefer to push this first.

pdunaj commented:
hi @carlescufi , can you merge it?

carlescufi commented:
@pdunaj I'd like @MaureenHelm to take a look at it first

carlescufi reviewed (approved):
Looks OK to me.

MaureenHelm commented on dts/bindings/sensor/nordic,nrf-qdec.yaml:
```diff
@@ -0,0 +1,74 @@
+
+    led-pin:
+      type: int
+      description: LED pin for light based QDEC device (0xFFFFFFFF if unused)
```
Instead of 0xFFFFFFFF if unused, make the category optional.

pdunaj replied:
Fixed

MaureenHelm commented on include/sensor.h:
```diff
@@ -143,6 +143,9 @@ enum sensor_channel {
 	SENSOR_CHAN_CURRENT,
 
+	/** Angular rotation, in degrees */
+	SENSOR_CHAN_ROTATION,
```
I wonder if angular position might be better. "Rotation" could mean position or velocity.

pdunaj replied:
I thought about it but I think angular position will be understood as absolute value whereas qdec gives relative change (+/- from previous reading).

MaureenHelm replied:
Doesn't that make it velocity then?

pdunaj replied:
Velocity is a change of distance in time. Here we have no idea how much time passed only that wheel rotated by `x` angle.

MaureenHelm commented on drivers/sensor/qdec_nrfx/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+zephyr_library()
```
Add license/copyright header to all new files, not just .c/.h

pdunaj replied:
Fixed

MaureenHelm commented on drivers/sensor/qdec_nrfx/qdec_nrfx.c:
```diff
@@ -0,0 +1,306 @@
+
+static void accumulate(struct qdec_nrfx_data *data, int16_t acc)
+{
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
```
Why are you using a spinlock?

pdunaj replied:
`data->acc` is accessed from two contexts

MaureenHelm replied:
Why not a semaphore, which is used in most other places?

carlescufi replied:
@pdunaj agree with @MaureenHelm here. We are not getting the kernel ready for SMP yet. At this point a semaphore is the lightweight locking mechanism that most of all  use for this sort of synchronization.

pdunaj replied:
@carlescufi - semaphore may sleep and this is called from the interrupt.
I could disable irqs but I see no point of it if we have splinlocks and can use them directly.

carlescufi replied:
Well, I do not even know if these are well-implemented on Arm yet. No other driver uses them as far as I know, so I would try to be conservative and simply use irq_lock/unlock

pdunaj replied:
Changed

If spinlocks are considered unsafe documentation should clearly say so or they should be removed altogether.
In our case I don't think they are however as for non smp cases they will be converted to irq_lock.

carlescufi replied:
@pdunaj I do agree with you, but I believe it's premature to use spinlocks at this point.

MaureenHelm commented on drivers/sensor/qdec_nrfx/qdec_nrfx.c:
```diff
@@ -0,0 +1,306 @@
+	struct qdec_nrfx_data *data = &qdec_nrfx_data;
+
+	int16_t acc;
+	int16_t accdbl;
```
These should be `s16_t`

pdunaj replied:
not really - these variables are arguments passed to `nrfx_qdec_accumulators_read` (both args are `int16_t`)

MaureenHelm replied:
Ok

MaureenHelm commented on drivers/sensor/qdec_nrfx/qdec_nrfx.c:
```diff
@@ -0,0 +1,306 @@
+	k_spin_unlock(&data->lock, key);
+
+	static_assert(CONFIG_QDEC_STEPS > 0, "only positive number valid");
+	static_assert(CONFIG_QDEC_STEPS <= 2148, "overflow possible");
```
use `__ASSERT` instead

pdunaj replied:
why adding check that runs every time (when enabled) instead of build time check?
this values does not change in runtime

MaureenHelm replied:
You're right, my mistake.

MaureenHelm commented on dts/bindings/sensor/nordic,nrf-qdec.yaml:
```diff
@@ -0,0 +1,74 @@
+      description: required interrupts
+      generation: define
+
+    a-pin:
```
Are these pins used as gpios? If so, they should have names like `a-gpios`, `b-gpios`, etc. to be consistent with other bindings.

MaureenHelm reviewed (changes requested):
Please add this new sensor driver to `tests/drivers/build_all/sensors_i_z.conf` to ensure it gets built in CI.

pdunaj commented:
Give me a second - I need to move change in bindings and fixup to a correct commit.

pdunaj committed:
drivers: sensors: qdec_nrfx: Add support for Nordic QDEC

Add a QDEC driver that utilizes Nordic's QDEC.

This closes issue #9364

Jira:DESK-259

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

pdunaj commented:
Fixes moved to correct commits

pdunaj commented:
Change tested on PCA20041 - everything seems to be working fine.

anangl commented on ext/hal/nordic/Kconfig:
```diff
@@ -26,6 +26,9 @@ config NRFX_PWM
 config NRFX_SAADC
 	bool
 
+config NRFX_QDEC
```
Keep the alphabetical order of the peripherals, please.

pdunaj replied:
Fixed

MaureenHelm commented:
Please create a follow up PR to convert to the new logger
